neonvm/controllers/virtualmachinemigration_controller.go

```diff
@@ -85,412 +86,543 @@ func (r *VirtualMachineMigrationReconciler) Reconcile(ctx context.Context, req c
+		migration.Status.Info.DowntimeMs = migrationInfo.DowntimeMs
+		migration.Status.Info.Ram.Transferred = migrationInfo.Ram.Transferred
+		migration.Status.Info.Ram.Remaining = migrationInfo.Ram.Remaining
+		migration.Status.Info.Ram.Total = migrationInfo.Ram.Total
+		migration.Status.Info.Compression.CompressedSize = migrationInfo.Compression.CompressedSize
+		migration.Status.Info.Compression.CompressionRate = int64(math.Round(migrationInfo.Compression.CompressionRate))
+		return r.updateMigrationStatus(ctx, migration)
+
 	case vmv1.VmmSucceeded:
-		// do nothing
+		// do additional VM status checks
+		if vm.Status.Phase == vmv1.VmMigrating {
+			// migration Succeeded and VM should have status Running
+			vm.Status.Phase = vmv1.VmRunning
+			// update VM status
+			if err := r.Status().Update(ctx, vm); err != nil {
+				log.Error(err, "Failed to update VM status from Migrating to Running as Migration succeeded")
+				return ctrl.Result{}, err
+			}
+		}
+
+		if len(migration.Status.SourcePodName) > 0 {
+			// try to find and remove source runner Pod
+			sourceRunner := &corev1.Pod{}
+			err := r.Get(ctx, types.NamespacedName{Name: migration.Status.SourcePodName, Namespace: migration.Namespace}, sourceRunner)
+			if err != nil && !apierrors.IsNotFound(err) {
+				log.Error(err, "Failed to get source runner Pod for deletion")
+				return ctrl.Result{}, err
+			}
+			if err := r.Delete(ctx, sourceRunner); err != nil {
+				log.Error(err, "Failed to delete source runner Pod")
+				return ctrl.Result{}, err
+			}
+			sourceRunnerPodName := migration.Status.SourcePodName
+			message := fmt.Sprintf("Source runner (%s) was deleted", sourceRunnerPodName)
+			log.Info(message)
+			r.Recorder.Event(migration, "Normal", "Deleted", message)
+			migration.Status.SourcePodName = ""
+			migration.Status.SourcePodIP = ""
+			return r.updateMigrationStatus(ctx, migration)
+		}
+		// all done, stop reconcilation
+		return ctrl.Result{}, nil
+
 	case vmv1.VmmFailed:
-		// do nothing
+		// do additional VM status checks
+		if vm.Status.Phase == vmv1.VmMigrating {
+			// migration Failed and VM should back to Running state
+			vm.Status.Phase = vmv1.VmRunning
+			if err := r.Status().Update(ctx, vm); err != nil {
+				log.Error(err, "Failed to update VM status from Migrating back to Running as Migration was failed")
+				return ctrl.Result{}, err
+			}
+		}
+		// all done, stop reconcilation
+		return ctrl.Result{}, nil
+
 	default:
-		// do nothing
+		// not sure what to do, so try rqueue
+		log.Info("Requeuing cuurent request")
+		return ctrl.Result{RequeueAfter: time.Second}, nil
+	}
+
+	// MAIN RECONCILE LOOP END
+	return ctrl.Result{}, nil
+}
+
+// finalizeVirtualMachineMigration will perform the required operations before delete the CR.
+func (r *VirtualMachineMigrationReconciler) updateMigrationStatus(ctx context.Context, migration *vmv1.VirtualMachineMigration, requeue ...bool) (ctrl.Result, error) {
```

Comment:
`updateMigrationStatus` is never called with `len(requeue) > 0`, is there any reason to keep it?

Reply:
it was used but... after several round of redesign these calls (`updateMigrationStatus(cdx, migration, true)`) disappeared :) Will remove this from function.

Reply:
removed